### PR TITLE
Add tutorial to formulate logistic regression with JuMP

### PR DESCRIPTION
**Pull request by frapac**
This is a short tutorial on the formulation of logistic regression problem with JuMP. 
It took its inspiration from [Mosek](https://docs.mosek.com/9.2/pythonfusion/case-studies-logistic.html), and present classical l2 regularized logistic regression, and logistic regression with l1 regularization (soft thresholding). We could include an extension to hard thresholding, but that would require a commercial solver to run...

I welcome any feedback!

### CODE REVIEW

File: script/modelling/logistic_regression.jl [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+#' title: Logistic Regression
+#' ---
+
+#' **Originally Contributed by**: François Pacaud
+
+#' This tutorial shows how to solve a logistic regression problem
+#' with JuMP. Logistic regression is a well known method in machine learning,
+#' useful when we want to classify binary variables with the help of
+#' a given set of features. Fitting a logistic
+#' regression problem sums up to find the optimal combination of features maximizing
```
Comment (matbesancon) [inline script/modelling/logistic_regression.jl]: "sums up to find" sounds odd here, is there a way to make the sentence simpler?

File: script/modelling/logistic_regression.jl [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+#' the (log)-likelihood onto a training set. In the point of view of optimization,
+#' the resulting problem is convex and differentiable. On a modern optimization
```
Comment (matbesancon) [inline script/modelling/logistic_regression.jl:13]: From a modern optimization glance?

File: script/modelling/logistic_regression.jl [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+#'
+#' Suppose we have a set of training data-point $i = 1, \cdots, n$, where
+#' for each $i$ we have a vector of features $x_i \in \mathbb{R}^p$ and a
+#' categorical observation $y_i \in \{-1, 1\}$.
+#'
+#' The log-likelihood is given by
+#' $$
+#' l(\theta) = \sum_{i=1}^n \log(\dfrac{1}{1 + \exp(-y_i \theta^\top x_i)})
+#' $$
+#' and finding the optimal parameter $\theta$ sums up to find the vector
```
Comment (matbesancon) [inline script/modelling/logistic_regression.jl]: > and finding the optimal parameter $\theta$ sums up to find the vector

and the optimal $\theta$ minimizes the logistic loss function

File: script/modelling/logistic_regression.jl [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+#' $\theta$ minimizing the logistic loss function:
+#' $$
+#' \min_{\theta}\; \sum_{i=1}^n \log(1 + \exp(-y_i \theta^\top x_i)) .
```
Comment (matbesancon) [inline script/modelling/logistic_regression.jl]:
```suggestion
#' \min_{\theta}\; \sum_{i=1}^n \log(1 + \exp(-y_i \theta^\top x_i)).
```

File: script/modelling/logistic_regression.jl [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+#'                    & \quad u_i = -y_i x_i^\top \theta \\
+#'                    & \quad r \geq \|\theta\|
+#' \end{aligned}
+#' $$
+#' and thus encompasses $3n + p + 1$ variables and $3n + 1$ constraints ($u_i = -y_i \theta^\top x_i$
+#' is only a temporary constraint used to clarify the notation).
+#' Thus, if $n \gg 1$, we get a large number of variables and constraints which
+#' could imped the resolution in the conic solver.
+
+#' ## Fitting logistic regression with a conic solver
```
Comment (matbesancon) [inline script/modelling/logistic_regression.jl:87]: I think these two sentences can be written in one

File: script/modelling/logistic_regression.jl [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+#' We start by implementing a function to generate a fake dataset, and where
+#' we could tune the correlation between the feature variables. The function
+#' is a direct transcription of the one used in [this blog post](http://fa.bianp.net/blog/2013/numerical-optimizers-for-logistic-regression/).
+function generate_dataset(n_samples=100, n_features=10; corr=0.0)
+    X = randn(n_samples, n_features)
+    w = randn(n_features)
+    y = sign.(X * w)
+    X .+= 0.8 * randn(n_samples, n_features) # add noise
+    X .+= corr # this makes it correlated by adding a constant term
```
Comment (matbesancon) [inline script/modelling/logistic_regression.jl]: I fail to see, how does a constant term adds correlation? Is it not just moving the data points by a constant vector in the feature space ?

If we need correlated features, should randn be called with a correlation matrix?

**Comment by matbesancon** [comment]
bump on the comments I left :)
also, there is a conflict on the runtests file

**Comment by frapac** [comment]
Oh, thanks for the reminder Mathieu! I will update the PR according to your comments in the week-end :)

File: script/modelling/logistic_regression.jl (hunk 6 above)
Comment (frapac) [reply script/modelling/logistic_regression.jl]: You are right about that, I should have been more careful when adapting the code from the origin blog post. I have renamed `corr` to `shift`, as we are only shifting the features by a constant value.

**Comment by matbesancon** [comment]
Hi @frapac sorry for the delay. The PR was not compatible with julia 1.0 which we are trying to maintain for now.
If you merge this one:
https://github.com/frapac/JuMPTutorials.jl/pull/1
It should be good

**Comment by matbesancon** [comment]
what a pain, it seems not every architecture is supported on MacOS

**Comment by frapac** [comment]
Maybe we could remove SCS in the deps. I remember that there was a discussion about that during the last JuMP-dev meeting, but I do not remember which solvers we should keep...

**Comment by frapac** [comment]
Would you like me to rebase this PR wrt master?

**Comment by matbesancon** [comment]
yes that would be perfect thanks, this should get rid of the conflict. You can also get rid of the manifest

**Comment by frapac** [comment]
Mmhh I made a mistake in the `converter/Manifest.toml` file... Is this file needed currently, or could we get rid of it? Otherwise, I will take the version of the Manifest on `master`

**Comment by matbesancon** [comment]
I think the converter/Manifest.toml also needs regenerating

**Comment by matbesancon** [comment]
I think this one can stay or be ignored, we fixed the necessary versions in the Project.toml, so you could remove it and add it to .gitignore

**Comment by odow** [comment]
Want to move this to the main JuMP documentation? You can find the tutorial here: https://github.com/jump-dev/JuMP.jl/tree/master/docs/src/tutorials

**Comment by frapac** [comment]
> Want to move this to the main JuMP documentation? 

With pleasure! I have transferred the PR here: https://github.com/jump-dev/JuMP.jl/pull/2499

**Pull request closed without merging**